### PR TITLE
refactor: add `Type` to interpolator class and factory method for usability

### PR DESCRIPTION
**Pull request by apaletta3**
I have refactor the `Interpolator` class and its children to follow the pattern of `Dynamics` and `GuidanceLaw` in ostk-astro to have a composable `InterpolationType` enum that can be used to figure out which interpolator type you're using under the hood. Part of the refactor involves properly using virtual functions and inheritance, as the previous version was not doing this correctly.

If you agree with this I'll add the python bindings just like they are done for `Dynamics` and `GuidanceLaw`

### CODE REVIEW

**Comment by vishwa2710** [comment]
Hey Antoine, thanks for working on this, great effort. This definitely improves the current paradigm by avoiding the interpolator class having to dynamically cast.

However, it's not really making it "composable". We're still relying on object inheritance to accomplish this using types.

A composable solution would leverage type-erasure using std::function or something like that. And TBD if that is the best solution for this.
This is still better than our current implementation.

One issue that still persists is,
To create an interpolator at runtime we will still have to do:
```
if (interpolator_type == "CubicSpline"){
   interpolator = std::make_shared<Interpolator>(CubicSpline...);
}
else if..
...
```
In any class that is instantiating an interpolator.

To alleviate this we could create a factory method that does something like:
```
std::unique_ptr<Interpolator> GenerateInterpolator(Interpolator::Type type) {
    switch (type) {
        case Interpolator::Type::Linear:
            return std::make_unique<Linear>();
        case Interpolator::Type::CubicSpline:
            return std::make_unique<CubicSpline>();
        // ... other types ...
        default:
            throw std::invalid_argument("Invalid interpolator type");
    }
}
```

Also, we can establish that any interpolator will required a VectorXd of x, and VectorXd of y to be instantiated. So we can make those members a part of the parent, simplifying the constructors for the children.

**Comment by kyle-cochran** [comment]
Nice dude awesome work. 

In addition to Vishwa's comment (which I agree with), another option is to use a [PIMPL](https://en.cppreference.com/w/cpp/language/pimpl) paradigm like we commonly do in Physics ([example for Earth Atmospheric model](https://github.com/open-space-collective/open-space-toolkit-physics/blob/main/src/OpenSpaceToolkit/Physics/Environment/Atmospheric/Earth.cpp)).

tl;dr for this one: instead of having a feature-full child interpolator `BarycentricRational` that inherits from an abstract parent `Interpolator`, you have an "interface" class `Interpolator` with a type and a **pointer** to the implementation logic:

```
class Interpolator {
 public:

  enum class Type {
    Linear,
    CubicSpline,
    ...
  }
 
 Result evaluate(...) ;

 private:
  Type _type ;
  Unique<InterpolatorImpl> _impl ;
}
```

You then have all of the disparate implementation logic inherit from `InterpolatorImpl`. You can have `InterpolatorImpl` define any required attributes or methods that an implementation must have, such as an `evaluate(...)` function. You define a pass-through function that calls the underlying function of the implementation:

```
Result Interpolator::evaluate(...) {
  return this->_impl->evaluate(...) ;
}
```

The main benefit here is that the class that a user must interact with now is _always_ `Interpolator` and you can have syntax like this:
```
Interpolator linearInterpolator = Interpolator(Interpolator::Type::Linear, <other args>) ;
Result result = linearInterpolator.evaluate(...) ;
```

This solution _is_ a bit more composable I would say, since you are pointing to different implementations behind the scenes based on a `Type` parameter. Although it's still not a purely composable solution.

The downside of this one is that it is probably more boilerplate. For every function that you want to expose for different types of Interpolators, you **must** add a pass-through function to the interface class `Interpolator`.

However, it does avoid the awkwardness of C++ object memory mapping by just hiding all of the inheritance behind a pointer.


I don't know if we use this paradigm in Astro anywhere, and I'm not necessarily advocating that we switch to it, but it is how we handle a similar problem in Phyics, so I thought I'd point it out.

**Comment by apaletta3** [comment]
Agreed, I didn't make the classes fully composable (just composable in terms of the InterpolatorType that each Interpolate has - which idk if you can really call that composable lol) because I didn't know how that would work when you need to interchangeably use one or the other like for Dynamics or GuidanceLaw and so just re-used those design patterns because the use case seemed similar.

I was trying to find a good way to not have to use switch statements to create the type of interpolator you want, and the GenerateInterpolator factor method does just that so thanks @vishwa2710 ! And yes I agree the Pimpl pattern would be the cleanest from the user perspective and I was aware of its use in OSTk-physics, but indeed with the disadvantage of extra code and complexity when it maybe isn't 100% needed. @kyle-cochran

**Comment by apaletta3** [comment]
Ready for final review!

Interesting other small detail with pybind, it seems that you don't actually have to define the pure virtual method of the parent class that is being overriden by a child class method in the regular binding (not the trampoline helper class), but it probably makes sense to define them anyways for documentation and consistency purposes. 

In case that didn't make sense^, I mean that it seems you can do this

```cpp
// Trampoline class for virtual member functions
class PyInterpolator : public Interpolator
{
   public:
    using Interpolator::Interpolator;

    // Trampoline (need one for each virtual function)

    VectorXd evaluate(const VectorXd& aQueryVector) const override
    {
        PYBIND11_OVERRIDE_PURE(VectorXd, Interpolator, evaluate, aQueryVector);
    }

    double evaluate(const double& aQueryValue) const override
    {
        PYBIND11_OVERRIDE_PURE(double, Interpolator, evaluate, aQueryValue);
    }
};

inline void OpenSpaceToolkitMathematicsPy_CurveFitting_Interpolator(pybind11::module& aModule)
{
    class_<Interpolator, PyInterpolator, Shared<Interpolator>> interpolator_class(aModule, "Interpolator");

    interpolator_class

        .def(init<const Interpolator::InterpolationType&>(), arg("interpolation_type"))

        .def("get_interpolation_type", &Interpolator::getInterpolationType)

        .def("evaluate", overload_cast<const VectorXd&>(&Interpolator::evaluate, const_), arg("x"))
        .def("evaluate", overload_cast<const double&>(&Interpolator::evaluate, const_), arg("x"))

        .def_static(
            "generate_interpolator", &Interpolator::GenerateInterpolator, arg("interpolation_type"), arg("x"), arg("y")
        )

        ;
```

or this 
```cpp

// Trampoline class for virtual member functions
class PyInterpolator : public Interpolator
{
   public:
    using Interpolator::Interpolator;

    // Trampoline (need one for each virtual function)

    VectorXd evaluate(const VectorXd& aQueryVector) const override
    {
        PYBIND11_OVERRIDE_PURE(VectorXd, Interpolator, evaluate, aQueryVector);
    }

    double evaluate(const double& aQueryValue) const override
    {
        PYBIND11_OVERRIDE_PURE(double, Interpolator, evaluate, aQueryValue);
    }
};

inline void OpenSpaceToolkitMathematicsPy_CurveFitting_Interpolator(pybind11::module& aModule)
{
    class_<Interpolator, PyInterpolator, Shared<Interpolator>> interpolator_class(aModule, "Interpolator");

        .def(init<const Interpolator::InterpolationType&>(), arg("interpolation_type"))

        .def("get_interpolation_type", &Interpolator::getInterpolationType)

		// REMOVED base class virtual meethods

        .def_static(
            "generate_interpolator", &Interpolator::GenerateInterpolator, arg("interpolation_type"), arg("x"), arg("y")
        )

        ;
```
I find this strange cause the [documentation](https://pybind11.readthedocs.io/en/stable/advanced/classes.html#overriding-virtual-functions-in-python) again is slightly vague around this subject, and seems to imply that you do need to define the virtual method of the parent class once in the trampoline class and once in the actual binding.

File: include/OpenSpaceToolkit/Mathematics/CurveFitting/Interpolator.hpp [hunk 1]
```diff
@@ -16,46 +18,60 @@ namespace curvefitting
...
+    /// @param anInterpolationType Interpolation type
+    /// @param anXVector A vector of x values
+    /// @param aYVector A vector of y values
+    /// @return Shared pointer to correct Interpolator
+    static const Shared<const Interpolator> GenerateInterpolator(
+        const InterpolationType& anInterpolationType, const VectorXd& anXVector, const VectorXd& aYVector
+    );
+
+   private:
+    InterpolationType interpolationType_;
```
Comment (vishwa2710) [inline include/OpenSpaceToolkit/Mathematics/CurveFitting/Interpolator.hpp:74]:
```suggestion
    InterpolationType interpolationType_;
    VectorXd x_;
    VectorXd y_;
```

File: include/OpenSpaceToolkit/Mathematics/CurveFitting/Interpolator.hpp [hunk 2]
```diff
@@ -16,46 +18,60 @@ namespace curvefitting
...
+    enum class InterpolationType
+    {
+        BarycentricRational,
+        CubicSpline,
+        Linear
+    };
 
-    /// @brief              Destructor (pure virtual)
+    /// @brief Constructor
+    Interpolator(const InterpolationType& anInterpolationType);
```
Comment (vishwa2710) [inline include/OpenSpaceToolkit/Mathematics/CurveFitting/Interpolator.hpp:42]:
```suggestion
    Interpolator(const InterpolationType& anInterpolationType, const VectorXd& anXVector, const VectorXd& aYVector);
```
Since these are used by every interpolator, we could make it a part of the parent class.
You can add getters as well.

File: include/OpenSpaceToolkit/Mathematics/CurveFitting/Interpolator.hpp [hunk 3]
```diff
@@ -16,46 +18,60 @@ namespace curvefitting
...
-    Interpolator() = default;
+    enum class InterpolationType
+    {
+        BarycentricRational,
+        CubicSpline,
+        Linear
+    };
 
-    /// @brief              Destructor (pure virtual)
+    /// @brief Constructor
```
Comment (vishwa2710) [inline include/OpenSpaceToolkit/Mathematics/CurveFitting/Interpolator.hpp:41]: Update the docstrings here 🙏

File: test/OpenSpaceToolkit/Mathematics/CurveFitting/Interpolator/BarycentricRational.test.cpp [hunk 4]
```diff
@@ -32,6 +38,19 @@ TEST(OpenSpaceToolkit_Mathematics_Interpolator_BarycentricRational, Constructor)
...
+    x << 0.0, 1.0, 2.0, 3.0, 4.0, 5.0;
+
+    VectorXd y(6);
+    y << 0.0, 3.0, 5.0, 6.0, 9.0, 15.0;
+
+    BarycentricRational interpolator = BarycentricRational(x, y);
+
+    EXPECT_EQ(BarycentricRational::InterpolationType::BarycentricRational, interpolator.getInterpolationType());
+}
+
```
Comment (vishwa2710) [inline test/OpenSpaceToolkit/Mathematics/CurveFitting/Interpolator/BarycentricRational.test.cpp:53]:
```suggestion
```
This test is unnecessary as it has been tested in the parent class.

File: test/OpenSpaceToolkit/Mathematics/CurveFitting/Interpolator/CubicSpline.test.cpp [hunk 5]
```diff
@@ -42,6 +48,19 @@ TEST(OpenSpaceToolkit_Mathematics_Interpolator_CubicSpline, SecondConstructor)
...
+    x << 0.0, 1.0, 2.0, 3.0, 4.0, 5.0;
+
+    VectorXd y(6);
+    y << 0.0, 3.0, 5.0, 6.0, 9.0, 15.0;
+
+    CubicSpline interpolator = CubicSpline(x, y);
+
+    EXPECT_EQ(CubicSpline::InterpolationType::CubicSpline, interpolator.getInterpolationType());
+}
+
```
Comment (vishwa2710) [inline test/OpenSpaceToolkit/Mathematics/CurveFitting/Interpolator/CubicSpline.test.cpp:63]:
```suggestion
```
This test is unnecessary as it has been tested in the parent class.

File: test/OpenSpaceToolkit/Mathematics/CurveFitting/Interpolator/Linear.test.cpp [hunk 6]
```diff
@@ -32,6 +38,19 @@ TEST(OpenSpaceToolkit_Mathematics_Interpolator_Linear, Constructor)
...
+    x << 0.0, 1.0, 2.0, 3.0, 4.0, 5.0;
+
+    VectorXd y(6);
+    y << 0.0, 3.0, 5.0, 6.0, 9.0, 15.0;
+
+    Linear interpolator = Linear(x, y);
+
+    EXPECT_EQ(Linear::InterpolationType::Linear, interpolator.getInterpolationType());
+}
+
```
Comment (vishwa2710) [inline test/OpenSpaceToolkit/Mathematics/CurveFitting/Interpolator/Linear.test.cpp:53]:
```suggestion
```
This test is unnecessary as it has been tested in the parent class.

File: bindings/python/test/curve_fitting/interpolator/test_linear.py [hunk 7]
```diff
@@ -1,11 +1,26 @@
...
-        Linear(x=[0.0, 1.0, 2.0, 4.0, 5.0, 6.0], y=[0.0, 3.0, 6.0, 9.0, 17.0, 5.0])
+    def test_constructors(self, interpolator: Linear):
+        assert interpolator is not None
+        assert isinstance(interpolator, Interpolator)
+        assert isinstance(interpolator, Linear)
+
+    def test_get_interpolation_type(self, interpolator: Linear):
+        assert (
+            interpolator.get_interpolation_type() == Interpolator.InterpolationType.Linear
+        )
```
Comment (vishwa2710) [inline bindings/python/test/curve_fitting/interpolator/test_linear.py:23]:
```suggestion
```
This test is unnecessary as it has been tested in the parent class.

File: bindings/python/test/curve_fitting/interpolator/test_interpolator.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        assert isinstance(interpolator, Interpolator)
+
+        interpolator: Interpolator = Interpolator.generate_interpolator(
+            Interpolator.InterpolationType.Linear,
+            x=[0.0, 1.0, 2.0, 4.0, 5.0, 6.0],
+            y=[0.0, 3.0, 6.0, 9.0, 17.0, 5.0],
+        )
+
+        assert interpolator is not None
+        assert isinstance(interpolator, Interpolator)
```
Comment (vishwa2710) [inline bindings/python/test/curve_fitting/interpolator/test_interpolator.py:63]:
```suggestion
@pytest.mark.parametrize("interpolation_type, x, y", [
    (Interpolator.InterpolationType.BarycentricRational, [0.0, 1.0, 2.0, 4.0, 5.0, 6.0], [0.0, 3.0, 6.0, 9.0, 17.0, 5.0]),
    (Interpolator.InterpolationType.CubicSpline, [0.0, 1.0, 2.0, 3.0, 4.0, 5.0], [0.0, 3.0, 6.0, 9.0, 17.0, 5.0]),
    (Interpolator.InterpolationType.Linear, [0.0, 1.0, 2.0, 4.0, 5.0, 6.0], [0.0, 3.0, 6.0, 9.0, 17.0, 5.0])
])
def test_generate_interpolators(self, interpolation_type: Interpolator.Type, x: list[float], y: list[float]):
    interpolator: Interpolator = Interpolator.generate_interpolator(
        interpolator_type=interpolator_type,
        x=x,
        y=y,
    )

    assert interpolator is not None
    assert isinstance(interpolator, Interpolator)
```

File: include/OpenSpaceToolkit/Mathematics/CurveFitting/Interpolator.hpp [hunk 9]
```diff
@@ -16,46 +18,60 @@ namespace curvefitting
...
-/// @ref                        https://en.wikipedia.org/wiki/Interpolator.
-
+/// @ref https://en.wikipedia.org/wiki/Interpolator.
 class Interpolator
 {
    public:
-    /// @brief              Default constructor
-    ///
-    Interpolator() = default;
+    enum class InterpolationType
```
Comment (vishwa2710) [inline include/OpenSpaceToolkit/Mathematics/CurveFitting/Interpolator.hpp:34]:
```suggestion
    enum class Type
```
Since it is namespaced with Interpolator we can just call it Type.

File: bindings/python/test/curve_fitting/interpolator/test_interpolator.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
+# Apache License 2.0
+
+import pytest
+
+from ostk.mathematics.curve_fitting import Interpolator
+
+
+@pytest.fixture
+def interpolation_type() -> Interpolator.InterpolationType:
```
Comment (vishwa2710) [inline bindings/python/test/curve_fitting/interpolator/test_interpolator.py:9]: Update accordingly for this all other cases of `Interpolator.InterpolationType` -> `Interpolator.Type`

File: bindings/python/test/curve_fitting/interpolator/test_cubic_spline.py [hunk 11]
```diff
@@ -4,16 +4,32 @@
...
 
     def test_default_constructor_2(self):
         CubicSpline(y=[0.0, 3.0, 6.0, 9.0, 17.0, 5.0], x_0=0.0, h=1.0)
 
+    def test_get_interpolation_type(self, interpolator: CubicSpline):
+        assert (
+            interpolator.get_interpolation_type()
+            == Interpolator.InterpolationType.CubicSpline
+        )
+
```
Comment (vishwa2710) [inline bindings/python/test/curve_fitting/interpolator/test_cubic_spline.py:32]:
```suggestion
```
Already tested in parent.

File: bindings/python/test/curve_fitting/interpolator/test_barycentric_rational.py [hunk 12]
```diff
@@ -1,27 +1,40 @@
...
+        assert isinstance(interpolator, Interpolator)
+        assert isinstance(interpolator, BarycentricRational)
 
-    def test_evaluate(self):
-        spline = BarycentricRational(
-            x=[0.0, 1.0, 2.0, 4.0, 5.0, 6.0], y=[0.0, 3.0, 6.0, 9.0, 17.0, 5.0]
+    def test_get_interpolation_type(self, interpolator: BarycentricRational):
+        assert (
+            interpolator.get_interpolation_type()
+            == Interpolator.InterpolationType.BarycentricRational
```
Comment (vishwa2710) [inline bindings/python/test/curve_fitting/interpolator/test_barycentric_rational.py:25]:
```suggestion
```
Tested in parent.

**Review by vishwa2710 — CHANGES_REQUESTED** [review]
Looks great man! Just some comments on naming and class parameters.

File: include/OpenSpaceToolkit/Mathematics/CurveFitting/Interpolator.hpp (hunk 3 above)
Comment (apaletta3) [reply include/OpenSpaceToolkit/Mathematics/CurveFitting/Interpolator.hpp:41]: but it can't actually be called by the user though, so it shouldn't have any code{.cpp} associated with it right?

Comment (vishwa2710) [reply include/OpenSpaceToolkit/Mathematics/CurveFitting/Interpolator.hpp:41]: Yeah you don't need to add `code` but just the parameters and return value would be nice.

File: include/OpenSpaceToolkit/Mathematics/CurveFitting/Interpolator.hpp (hunk 2 above)
Comment (apaletta3) [reply include/OpenSpaceToolkit/Mathematics/CurveFitting/Interpolator.hpp:42]: I see what you mean, the drawback here is that we increase the coupling between all the interpolators and the base class

Comment (vishwa2710) [reply include/OpenSpaceToolkit/Mathematics/CurveFitting/Interpolator.hpp]: Yeah that's a fair point about double storing the data.

**Review by vishwa2710 — APPROVED** [review]
